crates/noirc_frontend/src/hir/type_check/expr.rs

```diff
@@ -153,9 +153,12 @@ impl<'interner> TypeChecker<'interner> {
                 let start_span = self.interner.expr_span(&for_expr.start_range);
                 let end_span = self.interner.expr_span(&for_expr.end_range);
 
-                let mut unify_loop_range = |actual_type, span| {
+                let mut unify_loop_range = |actual_type: &Type, span: Span| {
                     let expected_type = if self.is_unconstrained() {
-                        Type::FieldElement(CompTime::new(self.interner))
+                        Type::PolymorphicInteger(
+                            CompTime::new(self.interner),
+                            Shared::new(TypeBinding::Bound(actual_type.clone())),
```

Comment:
Creating a new polymorphic integer here for the loop start and end would allow them to be different types. Ie. for i in 0u23 .. 7u64 would be allowed.

Also, if it is a design decision that we want to allow loops over integer variables, we should allow them in non-unconstrained functions as well. Assuming they are still comptime of course.

Reply:
> Blocking this on the fact that we would then have unconstrained functions have sligtly different loop syntax to normal functions

The syntax should be the same in either case

Reply:
Ah, another thing to consider is that if the loop range is (for example) a `u32`, then the loop variable `i` should likely be typed as a `u32` as well.

Reply:
>Also, if it is a design decision that we want to allow loops over integer variables, we should allow them in non-unconstrained functions as well. Assuming they are still comptime of course.

Yep --  this was why I was blocking this PR

Reply:
This is a bit roundabout in that it is making a new variable `a` then unifying `a = start`, then `b = end`, then `start = end`. We should change it to either `a = start; a = end` or `a = start; start = end;`

We can probably get rid of the closure now, and create the type once then do the `a = start; start = end` approach.